File: apollo-runtime/src/main/java/com/apollographql/android/cache/normalized/CacheJsonStreamReader.java
package com.apollographql.android.cache.normalized;

import com.apollographql.android.impl.JsonReader;
import com.apollographql.android.impl.ResponseJsonStreamReader;

import java.io.IOException;

/**
 * A {@link ResponseJsonStreamReader} which checks reads serialized {@link CacheReference} to strongly
 * typed {@link CacheReference} instances.
 */
final class CacheJsonStreamReader extends ResponseJsonStreamReader {

  CacheJsonStreamReader(JsonReader jsonReader) {
    super(jsonReader);
  }

  @Override protected Object readScalar(ResponseJsonStreamReader streamReader) throws IOException {
    Object scalar = super.readScalar(streamReader);
    if (scalar instanceof String) {
      String scalarString = (String) scalar;
      if (CacheReference.isSerializedReference((String) scalar)) {
        return CacheReference.deserialize(scalarString);
      }
    }
    return scalar;
  }

}

File: apollo-normalized-cache-api/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/internal/CacheJsonStreamReader.kt
package com.apollographql.apollo.cache.normalized.internal

import com.apollographql.apollo.api.internal.Throws
import com.apollographql.apollo.api.internal.json.JsonReader
import com.apollographql.apollo.api.internal.json.ResponseJsonStreamReader
import com.apollographql.apollo.cache.normalized.CacheReference
import okio.IOException

/**
 * A [ResponseJsonStreamReader] with additional support for [CacheReference].
 */
class CacheJsonStreamReader(jsonReader: JsonReader) : ResponseJsonStreamReader(jsonReader) {

  @Throws(IOException::class)
  override fun nextScalar(optional: Boolean): Any? {
    val scalar = super.nextScalar(optional)
    return if (scalar is String && CacheReference.canDeserialize(scalar)) {
      CacheReference.deserialize(scalar)
    } else {
      scalar
    }
  }
}
